core/persister/src/main/scala/org/apache/openwhisk/core/database/persister/ActivationConsumer.scala

```diff
@@ -0,0 +1,152 @@
+    val r = config.retry
+    val f = RestartSource
+      .onFailuresWithBackoff(
+        minBackoff = r.minBackoff,
+        maxBackoff = r.maxBackoff,
+        randomFactor = r.randomFactor,
+        maxRestarts = r.maxRestarts) { () =>
+        //TODO Metric - Restarts
+        logging.info(this, "Starting the Kafka consumer source")
+        Consumer
+          .committableSource(consumerSettings(), createSubscription())
+          .mapAsyncUnordered(config.parallelism) { msg =>
+            val f = Try(parseActivation(msg.record.value())) match {
+              case Success(Some((tid, a))) => persist(a)(tid)
+              case Success(None)           => Future.successful(Done)
+              case Failure(e) =>
+                logging.warn(this, s"Error parsing json for record ${msg.record.key()}" + e)
+                Future.successful(Done)
+            }
+            f.map(_ => msg.committableOffset)
+          }
+          .mapMaterializedValue(c => control.set(c))
+          .via(Committer.flow(committerDefaults))
+      }
+      .runWith(Sink.ignore)
+
+    f.failed.foreach(t => logging.error(this, "KafkaConsumer failed " + t.getMessage))
+    f
+  }
+
+  private def createSubscription() = {
+    if (config.topicIsPattern) {
+      Subscriptions
+        .topicPattern(config.topic)
+        .withRebalanceListener(rebalanceListener)
+    } else {
+      Subscriptions.topics(config.topic)
+    }
+  }
+
+  private val lagRecorder =
+    system.scheduler.schedule(10.seconds, 10.seconds)(queueMetric.gauge.set(consumerLag))
+
+  def shutdown(): Future[Done] = {
+    lagRecorder.cancel()
+    val f = control.get().drainAndShutdown(streamFuture)(system.dispatcher)
+    f.onComplete(_ => system.stop(rebalanceListener))
+    f
+  }
+
+  def consumerLag: Long = server.getAttribute(name, "records-lag-max").asInstanceOf[Double].toLong.max(0)
+
+  private def persist(act: WhiskActivation)(implicit tid: TransactionId): Future[Done] = {
+    //TODO Failure case handling - If there is an issue in storing then stream
+    persister.persist(act)
+  }
+
+  private def parseActivation(data: Array[Byte]): Option[(TransactionId, WhiskActivation)] = {
+    //TODO Metric - message size
+    //Avoid converting to string first and then to json. Instead directly parse to JSON
+    val js = JsonParser(ParserInput(data))
+    AcknowledegmentMessage.serdes.read(js) match {
+      case ResultMessage(tid, Right(wa))                            => Some((tid, wa))
+      case CombinedCompletionAndResultMessage(tid, Right(wa), _, _) => Some((tid, wa))
+      case _                                                        => None
+    }
+  }
+
+  private def consumerSettings(): ConsumerSettings[String, Array[Byte]] =
+    ConsumerSettings(system, new StringDeserializer, new ByteArrayDeserializer)
+      .withGroupId(config.groupId)
+      .withBootstrapServers(config.kafkaHosts)
+      .withProperty(ConsumerConfig.CLIENT_ID_CONFIG, config.clientId)
```

Comment:
Would it be worth to make this blank or to use the unique ID for each instance?
We may want to scale out the microservice at some point.
Then each instance will form one consumer group.

Reply:
Can be done. Currently `client.id` is used to access the lag stats from JMX MBean. Hence made it deterministic via config.

From [docs for client.id](http://kafka.apache.org/documentation.html#producerconfigs) I was not sure of its significance in multi consumer group settings. Currently if you do not specify the id then it gets something like `consumer-0` based on a static counter. So 2 consumer in same group would still end up getting same id.

> An id string to pass to the server when making requests. The purpose of this is to be able to track the source of requests beyond just ip/port by allowing a logical application name to be included in server-side request logging.